src/main.rs

```diff
@@ -1,92 +1,77 @@
-use futures::{
-    channel::mpsc::{channel, Receiver},
-    SinkExt, StreamExt,
-};
 use notify::{Event, RecommendedWatcher, RecursiveMode, Watcher};
-use std::{
-    panic,
-    path::{Path, PathBuf},
-};
-use std::{thread, time};
+use std::path::Path;
+use std::time;
+use tokio::sync::mpsc::{channel, Receiver};
 
 mod chmod;
 mod permissions;
 
 fn async_watcher() -> notify::Result<(RecommendedWatcher, Receiver<notify::Result<Event>>)> {
-    let (mut tx, rx) = channel(1);
+    let (tx, rx) = channel(512);
 
     // Automatically select the best implementation for your platform.
     // You can also access each implementation directly e.g. INotifyWatcher.
     let watcher = RecommendedWatcher::new(move |res| {
-        futures::executor::block_on(async {
-            match tx.send(res).await {
-                Ok(()) => {}
-                Err(err) => println!("watch error: {:?}", err),
-            };
-        })
+        match tx.blocking_send(res) {
+            Ok(()) => {}
+            Err(err) => println!("watch error: {:?}", err),
+        };
     })?;
 
     Ok((watcher, rx))
 }
 
-async fn async_watch(path: PathBuf) -> notify::Result<()> {
+async fn async_watch(path: &Path) -> notify::Result<()> {
     let (mut watcher, mut rx) = async_watcher()?;
 
     // Add a path to be watched. All files and directories at that path and
     // below will be monitored for changes.
     watcher.watch(path.as_ref(), RecursiveMode::Recursive)?;
 
-    thread::sleep(time::Duration::from_millis(500));
+    tokio::time::sleep(time::Duration::from_millis(500)).await;
+    chmod::update_permission_recursive(path);
 
-    chmod::update_permission_recursive(path.clone());
+    // Spawn the watcher in a new thread so it can't block the main thread.
+    let thread = tokio::spawn(async move {
+        while let Some(res) = rx.recv().await {
+            match res {
+                Ok(event) => {
+                    if event.kind.is_create() || event.kind.is_modify() || event.kind.is_other() {
+                        if cfg!(debug_assertions) {
+                            println!("watch event: {:?}", event.kind);
+                        }
 
-    while let Some(res) = rx.next().await {
-        match res {
-            Ok(event) => {
-                if event.kind.is_create() || event.kind.is_modify() || event.kind.is_other() {
-                    if cfg!(debug_assertions) {
-                        println!("watch event: {:?}", event.kind);
+                        permissions::check_permissions(event.paths);
                     }
-
-                    let paths: Vec<PathBuf> = event.clone().paths.clone();
-                    permissions::check_permissions(paths);
                 }
+                Err(err) => println!("watch error: {:?}", err),
             }
-            Err(err) => println!("watch error: {:?}", err),
         }
-    }
+    });
+
+    thread.await.ok();
```

Comment:
Shouldn't we handle this error? Or at least log it

Reply:
Yep, good point. This error would be thrown if the thread panics, I forgot to log it.